pkg/kubectl-nfd/dryrun.go

```diff
@@ -0,0 +1,199 @@
+
+	nfdv1alpha1 "sigs.k8s.io/node-feature-discovery/pkg/apis/nfd/v1alpha1"
+	"sigs.k8s.io/node-feature-discovery/pkg/apis/nfd/validate"
+)
+
+// Labels are a Kubernetes representation of discovered features.
+type Labels map[string]string
+
+// ExtendedResources are k8s extended resources which are created from discovered features.
+type ExtendedResources map[string]string
+
+// Annotations are used for NFD-related node metadata
+type Annotations map[string]string
+
+func DryRun(nodefeaturerulepath, nodeffeaturepath string) []error {
+	var validationErr []error
+	nfr := nfdv1alpha1.NodeFeatureRule{}
+	nf := nfdv1alpha1.NodeFeature{}
+
+	nfrFile, err := os.ReadFile(nodefeaturerulepath)
+	if err != nil {
+		return []error{fmt.Errorf("error reading NodeFeatureRule file: %v", err)}
+	}
+
+	err = yaml.Unmarshal(nfrFile, &nfr)
+	if err != nil {
+		return []error{fmt.Errorf("error reading NodeFeatureRule file: %v", err)}
+	}
+
+	nfFile, err := os.ReadFile(nodeffeaturepath)
+	if err != nil {
+		return []error{fmt.Errorf("error reading NodeFeatureRule file: %v", err)}
+	}
+
+	err = yaml.Unmarshal(nfFile, &nf)
+	if err != nil {
+		return []error{fmt.Errorf("error reading NodeFeatureRule file: %v", err)}
+	}
+
+	validationErr = append(validationErr, processNodeFeatureRule(nfr, nf)...)
+
+	return validationErr
+}
+
+func processNodeFeatureRule(nodeFeatureRule nfdv1alpha1.NodeFeatureRule, nodeFeature nfdv1alpha1.NodeFeature) []error {
+	var validationErr []error
+	var taints []corev1.Taint
+
+	extendedResources := make(map[string]string)
+	labels := make(map[string]string)
+	annotations := make(map[string]string)
+
+	for _, rule := range nodeFeatureRule.Spec.Rules {
+		fmt.Println("Processing rule: ", rule.Name)
+		ruleOut, err := rule.Execute(&nodeFeature.Spec.Features)
+		if err != nil {
+			validationErr = append(validationErr, fmt.Errorf("failed to process rule: %v - %v", rule.Name, err))
+			continue
+		}
+		// taints
+		taints = append(taints, ruleOut.Taints...)
+		// labels
+		for k, v := range ruleOut.Labels {
+			labels[k] = v
+		}
+		// extended resources
+		for k, v := range ruleOut.ExtendedResources {
+			// Dynamic Value
+			if strings.HasPrefix(v, "@") {
+				dvalue, err := getDynamicValue(v, &nodeFeature.Spec.Features)
+				if err != nil {
+					validationErr = append(validationErr, fmt.Errorf("failed to get dynamic value: %v", err))
+					continue
+				}
+				extendedResources[k] = dvalue
+				continue
+			}
+			extendedResources[k] = v
+		}
+		// annotations
+		for k, v := range ruleOut.Annotations {
+			annotations[k] = v
+		}
+	}
+
+	if len(taints) > 0 {
+		taintValidation := validate.Taints(taints)
+		fmt.Println("***\tTaints\t***")
+		for _, taint := range taints {
+			fmt.Println(taint)
+		}
+		if len(taintValidation) > 0 {
+			fmt.Println("\t-Validation errors-")
+			for _, err := range taintValidation {
+				fmt.Println(err)
+			}
+		}
+	}
+
+	if len(labels) > 0 {
+		labelValidation := validate.Labels(labels)
```

Comment:
There's one peculiarity related to the labels, annotations and extended resources: the default prefix will be appended to non-namespaced keys by nfd-master. I'm not entirely sure how this should be handled 🤔 Maybe we should move that part to the rule processing (so that the rule output will always have a prefix).

Reply:
marking as solved by #1455